docs/running-on-kubernetes.md

```diff
@@ -185,6 +185,21 @@ To use a secret through an environment variable use the following options to the
 --conf spark.kubernetes.executor.secretKeyRef.ENV_NAME=name:key
 ```
 
+## Pod Template
+Kubernetes allows defining pods from [template files](https://kubernetes.io/docs/concepts/workloads/pods/pod-overview/#pod-templates).
+Spark users can similarly use template files to define the driver or executor pod configurations that Spark configurations do not support.
+To do so, specify the spark properties `spark.kubernetes.driver.podTemplateFile` and `spark.kubernetes.executor.podTemplateFile`
+to point to local files accessible to the `spark-submit` process. To allow the driver pod access the executor pod template
+file, the file will be automatically mounted onto a volume in the driver pod when it's created.
+
+It is important to note that Spark is opinionated about certain pod configurations so there are values in the
+pod template that will always be overwritten by Spark. Therefore, users of this feature should note that specifying
+the pod template file only lets Spark start with a template pod instead of an empty pod during the pod-building process.
+For details, see the [full list](#pod-template-properties) of pod template values that will be overwritten by spark.
+
+Pod template files can also define multiple containers. In such cases, Spark will always assume that the first container in
+the list will be the driver or executor container.
```

Comment:
is it possible to use only extra containers and not Spark specific? Could we have a naming convention or a less error prone convention?

Reply:
This PR originally had an extra spark conf for container names, but we have decided to use the first container in the template instead. Users can have an empty first container in the pod spec template if they only want to add containers without changing Spark's executor or driver container

Reply:
Having a first container empty looks ugly to me.

Reply:
@skonto True, but this prevents the addition of a new spark conf for the container name